pkg/util/tags_builder.go

```diff
@@ -44,17 +61,171 @@ func (tb *TagsBuilder) Reset() {
+	hash []uint64
+}
+
+func newHashedTagsWithCapacity(cap int) hashedTags {
+	return hashedTags{
+		data: make([]string, 0, cap),
+		hash: make([]uint64, 0, cap),
+	}
+}
+
+func newHashedTagsFromSlice(tags []string) hashedTags {
+	hash := make([]uint64, 0, len(tags))
+	for _, t := range tags {
+		hash = append(hash, murmur3.StringSum64(t))
+	}
+	return hashedTags{
+		data: tags,
+		hash: hash,
+	}
+}
+
+func (h hashedTags) slice(i, j int) hashedTags {
+	return hashedTags{
+		data: h.data[i:j],
+		hash: h.hash[i:j],
+	}
+}
+
+func (h hashedTags) copy() []string {
+	return append(make([]string, 0, len(h.data)), h.data...)
+}
+
+func (h hashedTags) len() int {
+	return len(h.data)
+}
+
+// HashingTagsBuilder allows to build a slice of tags to generate the context while
+// reusing the same internal slice.
+type HashingTagsBuilder struct {
+	hashedTags
+}
+
+// NewHashingTagsBuilder returns a new empty TagsBuilder.
+func NewHashingTagsBuilder() *HashingTagsBuilder {
+	return &HashingTagsBuilder{
+		hashedTags: newHashedTagsWithCapacity(128),
+	}
+}
+
+// NewHashingTagsBuilderFromSlice return a new TagsBuilder with the input slice for
+// it's internal buffer.
+func NewHashingTagsBuilderFromSlice(tags []string) *HashingTagsBuilder {
+	return &HashingTagsBuilder{newHashedTagsFromSlice(tags)}
+}
+
+// Append appends tags to the builder
+func (tb *HashingTagsBuilder) Append(tags ...string) {
+	for _, t := range tags {
+		tb.data = append(tb.data, t)
+		tb.hash = append(tb.hash, murmur3.StringSum64(t))
+	}
+}
+
+// AppendHashed appends tags and corresponding hashes to the builder
+func (tb *HashingTagsBuilder) AppendHashed(src *HashedTags) {
+	tb.data = append(tb.data, src.data...)
+	tb.hash = append(tb.hash, src.hash...)
+}
+
+// SortUniq sorts and remove duplicate in place
+func (tb *HashingTagsBuilder) SortUniq() {
+	if tb.Len() < 2 {
+		return
+	}
+
+	sort.Sort(tb)
+
+	j := 0
+	for i := 1; i < len(tb.data); i++ {
+		if tb.hash[i] == tb.hash[j] && tb.data[i] == tb.data[j] {
+			continue
+		}
+		j++
+		tb.data[j] = tb.data[i]
+		tb.hash[j] = tb.hash[i]
+	}
+
+	tb.Truncate(j + 1)
+}
+
+// Reset resets the size of the builder to 0 without discaring the internal
+// buffer
+func (tb *HashingTagsBuilder) Reset() {
+	// we keep the internal buffer but reset size
+	tb.data = tb.data[0:0]
+	tb.hash = tb.hash[0:0]
+}
+
 // Truncate retains first n tags in the buffer without discarding the internal buffer
-func (tb *TagsBuilder) Truncate(len int) {
+func (tb *HashingTagsBuilder) Truncate(len int) {
 	tb.data = tb.data[0:len]
+	tb.hash = tb.hash[0:len]
 }
 
 // Get returns the internal slice
-func (tb *TagsBuilder) Get() []string {
+func (tb *HashingTagsBuilder) Get() []string {
 	return tb.data
 }
 
+// Hashes returns the internal slice of tag hashes
+func (tb *HashingTagsBuilder) Hashes() []uint64 {
+	return tb.hash
+}
+
 // Copy makes a copy of the internal slice
-func (tb *TagsBuilder) Copy() []string {
+func (tb *HashingTagsBuilder) Copy() []string {
```

Comment:
It might make sense to have a `CopyAsHashedTags` that takes a "snapshot" of the builder as an immutable value.  This might be useful in `context_resolver.go` if my comment on that file makes sense.

It might also make sense to have a `HashedTags()` method that "converts" a `HashingTagsBuilder` into a `HashedTags` instance (invalidating or resetting the builder), although I don't actually see anywhere that conversion is useful.